src/bentoml/_internal/utils/__init__.py

```diff
@@ -273,6 +273,39 @@ def copy_file_to_fs_folder(
     fs.copy.copy_file(src_fs, file_name, dst_fs, dst_path)
 
 
+def download_and_zip_git_repo(
+    url: str, ref: str, subdirectory: str | None, dst_path: str
+) -> str:
+    """
+    Download the git repo at the given url and ref, and zip it to a file at dst_path
+    """
+    import shutil
+    import subprocess
+
+    from ...exceptions import BentoMLException
+
+    name = os.path.splitext(os.path.basename(url))[0]
+
+    with fs.open_fs("temp://") as temp_fs:
+        dest_dir = temp_fs.getsyspath(name)
+        git_command = ["git", "clone", "--filter=blob:none", "--quiet", url, dest_dir]
+        env = {**os.environ, "GIT_TERMINAL_PROMPT": "0"}
+        try:
+            subprocess.check_call(git_command, env=env)
+            if ref:
+                subprocess.check_call(["git", "fetch", "-q", url, ref], cwd=dest_dir)
+                subprocess.check_call(["git", "checkout", "FETCH_HEAD"], cwd=dest_dir)
+        except subprocess.CalledProcessError as e:
+            raise BentoMLException(
+                f"Failed to clone git repository {url}: {e.stderr}"
+            ) from e
+        zipball = os.path.join(dst_path, subdirectory or name)
+        shutil.rmtree(os.path.join(dest_dir, ".git"), ignore_errors=True)
```

Comment:
should we just do a shadow clone? or even a sparse checkout instead?

Reply:
This piece of code is being used in both pip and unearth.

I think the reason is we do not know whether user has specified the `@ref` part, in that case, we have to clone the full history and checkout to it later. the `ref` is an arbitrary legal git reference so we can't pass it in `--branch` option with `git clone`.

We could do a shallow clone as a shorcut when `ref` isn't given.

Reply:
I see. Is there a current use case for this?

Reply:
The following git url is allowed:
```
git+ssh://git@github.com/pallets/flask.git@dev
```